src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -190,6 +191,16 @@ CitusBeginScan(CustomScanState *node, EState *estate, int eflags)
 		CitusBeginModifyScan(node, estate, eflags);
 	}
 
+	Job *workerJob = scanState->distributedPlan->workerJob;
+
+	if (!IsShardKeyValueAllowed(workerJob->partitionKeyValue))
```

Comment:
We missed something here: We should also check the co-location ID. Otherwise, we could be inserting the same partition key value into another distributed table which is not co-located and therefore might be on a different node.